Long drush call causes an infinite loop, when not using ingest sets.

whikloj opened the pull request:
The drupal timers are not reset when one is "started", so once you have used up 2/3 of the lock timeout. 

It tries to re-schedule it, except because we are re-using the same lock name then the first check still sees the previous runs time. 

So you are already over the 2/3rds of lock timeout. Rinse, repeat, ad infinitum. This just clears the time when you re-schedule.

Commit:
Long drush call causes an infinite loop.

adam-vessey commented:
This doesn't make sense... Are you adding other parameters to the batch, like setting `progressive` to `FALSE`? `$timers` is a global variable, so it shouldn't be maintained between different processes.

adam-vessey commented:
(Oops...)

whikloj commented:
I am using the uom_newspaper_batch module to load newspapers. I have made minor modifications due to directory layout with new scans but I am not adding new parameters to the batch. 

I admit I can't explain why it is failing for me. But it is and this did fix it.

adam-vessey commented:
You really have to dig into it to see it, but generally, each pass through the batch operation should happen in its own process\* when `progressive` is `TRUE` as we expect it (and is the default); therefore, all [Drupal timers](https://api.drupal.org/api/drupal/developer!globals.php/global/timers/7) should get reset... something we in fact have to [account for when tracking the total time](https://github.com/uml-digitalinitiatives/islandora_batch/blob/7.x-fix-loop/includes/ingest.batch.inc#L201-L205)...  That said, this "fix" should not be necessary.

\* It actually uses some funky conditions, creating a new process after some time or process memory usage threshold... 1 second and 128 MB come to mind, but I'd have to dig through code to be sure.

whikloj commented:
Okay, so if this is our [preprocessor](https://github.com/uml-digitalinitiatives/uofm_newspaper_batch/blob/7.x/uofm_newspaper_batch.drush.inc#L53-L78), then I could assume that progressive is true. Any other idea why I would get stuck in an infinite loop at [this time check](https://github.com/Islandora/islandora_batch/blob/7.x/includes/ingest.batch.inc#L398)?

adam-vessey commented:
Preprocessors and the `progressive` flag are completely separate... It would be present in [the code setting up the Drupal batch](https://github.com/Islandora/islandora_batch/blob/7.x/islandora_batch.module#L219-L235), or possibly an implementation of [`hook_batch_alter()`](https://api.drupal.org/api/drupal/modules!system!system.api.php/function/hook_batch_alter/7) in some module (or theme?) you have enabled.

Not sure that it makes sense to indicate that condition as the "cause" of an infinite loop... If it were that loop going indefinitely, the change in this pull request would never get run? At least, not until one of the other conditions fail?

whikloj commented:
Sorry guess I wasn't clear. What I seemed to be experiencing was that the script would run until that time check failed (ie. 2/3 of the lock timeout had expired). The while loop is not run, and instead the function [exits here](https://github.com/Islandora/islandora_batch/blob/7.x/includes/ingest.batch.inc#L474). 

It would fall down hit the [else](https://github.com/Islandora/islandora_batch/blob/7.x/includes/ingest.batch.inc#L259-L267) and update $context['sandbox']['loops'] and $context['finished']

Then you [store the timer](https://github.com/Islandora/islandora_batch/blob/7.x/includes/ingest.batch.inc#L269)

The next run of islandora_batch_ingest_process() would start per normal, everything would be good. 

Except that when the islandora_batch_process_results() was run, the timer still had "time" on it. It was not a brand new instance. So it would fail immediately and the cycle would start again.

I don't see where the time would be reset, timer_stop unsets the "start" index. So I decided to reset the "time".

So where in islandora_batch would be a good place to drop a watchdog to check for the progressive option?

adam-vessey commented:
Pay attention to the constants... `ISLANDORA_BATCH_TIMER_NAME` is getting stashed, and you're unsetting `ISLANDORA_BATCH_LOCK_TIMER`... These track different things. The first tracks the time for the entire batch (to support the timeout parameter), the second tracks _just_ the age of the lock, so we do not end up processing objects when we are not in possession of the lock.

Kinda thinking perhaps a detailed bug report (parameters being passed, modules in question, possibly a list of all modules installed...?... I'm not sure we really have a full list of things which would be useful) should be thrown onto a JIRA ticket, to be able to reproduce in order to identify the actual issue... I suspect an issue like this would also effect [processing individual sets](https://github.com/Islandora/islandora_batch/blob/7.x/includes/ingest.batch.inc#L253-L258).

whikloj commented:
Okay now I am confused too, I was sure when I fixed this it was the ISLANDORA_BATCH_TIMER_NAME. 

Once my current run is finished, I'm going to back out this change and see if I am dreaming or if I updated something else.

Regarding ingest_sets I am having a problem with them. Which is why I went back to pure batches and thought I should put in this PR. 

Which I will close. cheers.
